### DISCUSSION TITLE
REDIS_URL - Connection with username-password pair - fail

### DISCUSSION THREAD

**Discussion by bytesnackscgn** [outline/outline#5459] [Self Hosting]
I am using docker compose to create docker instances of outline.
If using environment variable  REDIS_URL with an username-password pair the connection is failing with 
ReplyError: WRONGPASS invalid username-password pair or user is disabled.
at parseError (/opt/outline/node_modules/redis-parser/lib/parser.js:179:12)
at parseType (/opt/outline/node_modules/redis-parser/lib/parser.js:302:14)
{"level":"warn","message":"Retrying redis connection: attempt 1"}

If using only password in REDIS_URL it returns error like "username password pair is may not activated or worng..."

If i try to use ioredis and convert a configuration like the following to base64 it also fails.
{ "host": "myip "port": myport, "password": "mypassword, "username": "myuser"}

Error report with base64 ioredis
{"error":"WRONGPASS invalid username-password pair or user is disabled.","level":"error","message":"Redis error in socketio adapter","stack":"ReplyError: WRONGPASS invalid username-password pair or user is disabled.\n    at parseError (/opt/outline/node_modules/redis-parser/lib/parser.js:179:12)\n    at parseType (/opt/outline/node_modules/redis-parser/lib/parser.js:302:14)"}
{"level":"warn","message":"Retrying redis connection: attempt 1"}
{"level":"warn","message":"Retrying redis connection: attempt 1"}
{"level":"warn","message":"Retrying redis connection: attempt 1"}
{"error":"WRONGPASS invalid username-password pair or user is disabled.","level":"error","message":"Redis error in socketio adapter","stack":"ReplyError: WRONGPASS invalid username-password pair or user is disabled.\n    at parseError (/opt/outline/node_modules/redis-parser/lib/parser.js:179:12)\n    at parseType (/opt/outline/node_modules/redis-parser/lib/parser.js:302:14)"}
{"level":"warn","message":"Retrying redis connection: attempt 1"}
{"level":"warn","message":"Retrying redis connection: attempt 1"}
{"level":"warn","message":"Retrying redis connection: attempt 1"}
{"level":"warn","message":"Retrying redis connection: attempt 1"}
{"error":"write EPIPE","level":"error","message":"Redis error in socketio adapter","stack":"Error: write EPIPE\n    at afterWriteDispatched (node:internal/stream_base_commons:160:15)\n    at writeGeneric (node:internal/stream_base_commons:151:3)\n    at Socket._writeGeneric (node:net:930:11)\n    at Socket._write (node:net:942:8)\n    at writeOrBuffer (node:internal/streams/writable:392:12)\n    at _write (node:internal/streams/writable:333:10)\n    at Writable.write (node:internal/streams/writable:337:10)\n    at RedisAdapter.sendCommand (/opt/outline/node_modules/ioredis/built/Redis.js:394:29)\n    at RedisAdapter.sendCommand (/opt/outline/node_modules/dd-trace/packages/datadog-instrumentations/src/ioredis.js:16:53)\n    at RedisAdapter.client (/opt/outline/node_modules/ioredis/built/utils/Commander.js:90:25)\n    at /opt/outline/node_modules/ioredis/built/redis/event_handler.js:224:18\n    at Socket.<anonymous> (/opt/outline/node_modules/ioredis/built/redis/event_handler.js:47:39)\n    at Object.onceWrapper (node:events:627:28)\n    at Socket.emit (node:events:525:35)\n    at Socket.emit (node:domain:489:12)\n    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1479:10)"}
{"level":"warn","message":"Retrying redis connection: attempt 1"}
{"level":"warn","message":"Retrying redis connection: attempt 1"}
{"level":"warn","message":"Retrying redis connection: attempt 2"}
{"error":"connect ECONNREFUSED IP:PORT","level":"error","message":"Redis error in socketio adapter","stack":"Error: connect ECONNREFUSED IP:PORT\n    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1494:16)"}

Passwort and username are "simple" but long. Alphanumeric and minus or underscore to avoid problems on parsing connectionurls.

I am using external redis and postgres. Therefore my docker compose file looks like this.

version: "3"
services:

  outline:
    container_name: outline
    image: docker.getoutline.com/outlinewiki/outline:latest
    env_file: ./.env
    ports:
      - "3003:3000"
    depends_on:
      - storage

  storage:
    container_name: minio
    image: minio/minio
    env_file: ./.env
    ports:
      - "9000:9000"
    entrypoint: sh
    command: -c 'minio server /data'
    deploy:
      restart_policy:
        condition: on-failure
    volumes:
      - ./data/storage/:/data
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9000/minio/health/live"]
      interval: 30s
      timeout: 20s
      retries: 3

volumes:
  storage-data:

**To Reproduce**
use .env
REDIS_URL=redis://redisuser:redispass@host:prot

**Expected behavior**
THAT IT WORKS

**Outline:**
- Install: self hosted
- Version: latest docker image

**Comment by thatguyatgithub** [comment]
I can confirm it happens to me as well. the config containing the redis connection pair is correct.

**Comment by lenaxia** [comment]
Can confirm this as well. It was working in 0.67.2 and broke somewhere along the way (not sure exactly when), possibly around 0.70 or so. I am able to attach to the pod and perform an `nc` command to demonstrate that the pod is able to reach redis, as well launching a redis debug container can connect and query my redis instance.

**Comment by lrstanley** [comment]
Running into the same on the latest version. Not using auth works fine, though. @tommoor is there a reason this was turned into a discussion?

**Reply by tommoor** [reply]
Issues related to self hosted configuration are converted to discussions generally unless there is a bug to fix.

Here the error is clear, and coming from Redis - Outline just passes the configured user and pass through.

**Comment by lenaxia** [comment]
here is the solution I found.

You need to use the ioredis:// protocol as opposed to redis. Specifically looking at how redis.ts is defined https://github.com/outline/outline/blob/a7dd5c6798d04fa96b589f95d624aa36037ad287/server/storage/redis.ts

it assumes anything proceeded by ioredis:// is base64 encoded. 

What you need to encode is this:

```

{
  "sentinels": [
    { "host": "redis-lb.databases.svc.cluster.local", "port": 26379 }
  ],
  "name": "redis-master",
  "password": "redis-auth-password",
  "sentinelPassword": "sentinel-auth-password"
}

```

Specifically here I am using port 26379 which is my sentinel port. My setup uses an HA redis with HA sentinels, hidden behind a HAproxy load balancer. So thats why there's only one entry here. If you dont have an LB, then have 3 entries for sentinels for each of your instances. You also need to find your sentinel instance name, mine was redis-master as found by connecting to one of your sentinels and running INFO SENTINEL

Encode it as such:

`echo -n '{"sentinels":[{"host":"redis-lb.databases.svc.cluster.local","port":26379}],"name":"redis-master","password":"redis-auth-password","sentinelPassword": "sentinel-auth-password"}' | base64`

Then set this in your environment like so:

```

      - name: REDIS_URL
        value: "ioredis://<YOUR BASE64 ENCODED STRING>"

```
        
Now I am no longer getting redis connection errors in my logs.

**Reply by lrstanley** [reply]
~~Going to test this out later today. Working on an updated/improved helm chart for kubernetes, so hoping I can figure out how to do all of that logic directly in the chart.~~

**update**: for non-sentinel usecases, this is the object I used:

```json
{"host": "my-hostname", "port": 1234, "password": "my-password"}
```

And of course, formatted as `ioredis://<json-as-base64>`
